### DISCUSSION TITLE
Issue P1001 - Can't reach database server (not always, only sometimes)

### DISCUSSION THREAD

**Discussion by yosef-cupist** [prisma/prisma#16816] [Q&A] [answered]
Good morning everyone, I will be very appreciated if anyone can give me advice or show me what is wrong with this below issue.

Currently, we are developing a web app using Prisma client to connect to AWS PostgreSQL. Almost of the time, it worked properly, sometimes it raised P1001 - Can't reach database server ...
- Environment:
  - Our web application is running under EKS.
  - Same VPC with AWS RDS
  - Prisma version: 4.5.0
- We have 2 Prisma clients:
  - writer instance: using default PrismaClient
    - DATABASE_URL: postgresql://user1:pass1@write-rds.xxxxx.rds.amazonaws.com:5432/enfpy
  - reader instance: initialized when creating PrismaService
    - READER_DATABASE_URL: postgresql://user1:pass1@read-only-rds.xxxxx.rds.amazonaws.com:5432/enfpy
  - The writer and reader instances are in the same rds cluster (same specs, and same VPC)
- Situation:
  - Our EKS and network have no problem at the time error occurred.
  - For the writer instance, this error occurred 1 time/week
     <img width="893" alt="Screenshot 2022-12-15 at 11 54 47" src="https://user-images.githubusercontent.com/97009052/207761613-1e0b1836-3922-4c18-95e7-82ca4f880604.png">
  - For the reader instance, error occurrent many times/day
    <img width="1687" alt="Screenshot 2022-12-15 at 11 54 02" src="https://user-images.githubusercontent.com/97009052/207761530-ae33f9ed-e2c2-4db4-a7db-4cd4e1275426.png">
- Error:
```
err.clientVersion: 4.5.0

err.code: P1001

err.message: 
Invalid `prisma.$queryRaw()` invocation:
Can't reach database server at `read-only-xxxxx.ap-northeast-2.rds.amazonaws.com`:`5432`

err.stack:
Error: 
Invalid `prisma.$queryRaw()` invocation:


Can't reach database server at `read-only-xxxx.ap-northeast-2.rds.amazonaws.com`:`5432`

Please make sure your database server is running at `read-only-xxxx.ap-northeast-2.rds.amazonaws.com`:`5432`.
at RequestHandler.handleRequestError (/usr/src/app/node_modules/@prisma/client/runtime/index.js:30873:13)
at RequestHandler.request (/usr/src/app/node_modules/@prisma/client/runtime/index.js:30856:12)
at async Proxy._request (/usr/src/app/node_modules/@prisma/client/runtime/index.js:31836:16)
at async getXXXXData (/usr/src/app/dist/apps/enfpy/main.js:1317:31)
...
```

- Our schema
```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["interactiveTransactions"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
```

- Our prisma client implementation
```
class PrismaService
  extends PrismaClient
  implements OnModuleInit, OnModuleDestroy
{
  readerClient: PrismaClient;

  constructor() {
    super();

    this.readerClient = new PrismaClient({
      datasources: {
        db: {
          url: getServerConfig().READER_DATABASE_URL,
        },
      },
    });
  }

  async onModuleInit() {
    await this.$connect();
    await this.readerClient.$connect();
  }

  async onModuleDestroy() {
    await this.$disconnect();
    await this.readerClient.$disconnect();
  }

  async enableShutdownHooks(app: INestApplication) {
    this.$on('beforeExit', async () => {
      await app.close();
    });
  }
}
const prismaService = new PrismaService();

export default prismaService;
```

**Comment by janpio** [comment, marked as answer]
Have you tried increasing the connection timeout values? `Can't reach ...` is also output when the connection take to long for any reason.
How many % of requests fail like this?

**Reply by yosef-cupist** [reply]
@janpio 

> Have you tried increasing the connection timeout values? Can't reach ... is also output when the connection take to long for any reason.

Mean put `?connect_timeout=30` in the connection string? If it was right, I already put this value to 30 seconds

> How many % of requests fail like this?

As I said above, it is very rare, under 0.001% of requests/day

**Reply by janpio** [reply]
> Mean put `?connect_timeout=30` in the connection string? If it was right, I already put this value to 30 seconds

Yes. Can you try a higher value? Most of the time this should of course not take that long - but it might be worth testing if it for some weird reason in your app sometimes, very rarely wants to take 31s and it failing because of that.

Generally any network based architecture will have some hickups per day. Could you maybe modify your app in a way that it "just" retries the query?

**Reply by yosef-cupist** [reply]
The situation didn't happen within a week after I increased `connect_timeout` to a higher value. But I still don't know the reason for this, but I think it is not from Prisma. 
Thank you and I will mark your question as answered.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
